### PR TITLE
Analyzer: compute ALIAS columns right after reading

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 8891fb237e4b27a377ad25bbb88bb2d13efc11c9 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59595/8891fb237e4b27a377ad25bbb88bb2d13efc11c9/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59595/8891fb237e4b27a377ad25bbb88bb2d13efc11c9/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by novikd** [comment]
I need this PR for #50894

**Comment by vdimir** [comment]
> Integration tests (asan, analyzer) [2/6] — fail: 5, passed: 358 [Details](https://s3.amazonaws.com/clickhouse-test-reports/59595/966fa3f44f32dcc26bcd97fb75adbf48bdac1f65/integration_tests__asan__analyzer__[2_6].html)

fixing row policy ...

File: src/Planner/TableExpressionData.h [hunk 1]
```diff
@@ -52,45 +52,48 @@ class TableExpressionData
...
     }
 
-    /// Add alias column name
-    void addAliasColumnName(const std::string & column_name, const ColumnIdentifier & column_identifier)
+    /// Mark existing column as selected
+    void markSelectedColumn(const std::string & column_name)
     {
-        alias_columns_names.insert(column_name);
+        auto [_, insetred] = selected_column_names_set.emplace(column_name);
+        if (insetred)
```
Comment (novikd) [inline src/Planner/TableExpressionData.h:83]:
```suggestion
        auto [_, inserted] = selected_column_names_set.emplace(column_name);
        if (inserted)
```

File: src/Planner/CollectTableExpressionData.cpp [hunk 2]
```diff
@@ -324,9 +347,10 @@ void collectTableExpressionData(QueryTreeNodePtr & query_node, PlannerContextPtr
     }
 }
 
-void collectSourceColumns(QueryTreeNodePtr & expression_node, PlannerContextPtr & planner_context)
+void collectSourceColumns(QueryTreeNodePtr & expression_node, PlannerContextPtr & planner_context, bool keep_alias_columns)
 {
-    CollectSourceColumnsVisitor collect_source_columns_visitor(*planner_context);
+    CollectSourceColumnsVisitor collect_source_columns_visitor(planner_context);
+    collect_source_columns_visitor.setKeepAliasColumns(keep_alias_columns);
```
Comment (novikd) [inline src/Planner/CollectTableExpressionData.cpp:353]: Can be a constructor argument

File: src/Planner/CollectTableExpressionData.cpp [hunk 3]
```diff
@@ -102,45 +114,51 @@ class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitor<CollectSource
...
 
-    PlannerContext & planner_context;
+private:
+    PlannerContextPtr & planner_context;
+
+    /// Replace ALIAS columns with their expressions or register them in table expression data.
+    /// Usually we can replace them when we build some "local" actions DAG
+    /// (for example Row Policy or PREWHERE) that is applied on right top of table expression.
+    bool keep_alias_columns = true;
+    bool select_added_columns = true;
```
Comment (novikd) [inline src/Planner/CollectTableExpressionData.cpp:161]: Add a comment

File: src/Planner/CollectTableExpressionData.cpp [hunk 4]
```diff
@@ -102,45 +114,51 @@ class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitor<CollectSource
...
     }
 
-    PlannerContext & planner_context;
+private:
+    PlannerContextPtr & planner_context;
+
+    /// Replace ALIAS columns with their expressions or register them in table expression data.
+    /// Usually we can replace them when we build some "local" actions DAG
+    /// (for example Row Policy or PREWHERE) that is applied on right top of table expression.
+    bool keep_alias_columns = true;
```
Comment (novikd) [inline src/Planner/CollectTableExpressionData.cpp:160]: When can not we do it?

**Comment by novikd** [comment]
Is it possible to remove alias columns computation in StorageMerge?

File: src/Storages/StorageDistributed.cpp [hunk 5]
```diff
@@ -744,6 +744,22 @@ StorageSnapshotPtr StorageDistributed::getStorageSnapshotForQuery(
...
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * column_node = node->as<ColumnNode>();
+        if (!column_node || !column_node->hasExpression())
+            return;
+        const auto & column_source = column_node->getColumnSourceOrNull();
+        if (column_source && (column_source->getNodeType() == QueryTreeNodeType::ARRAY_JOIN || column_source->getNodeType() == QueryTreeNodeType::JOIN))
+            return;
+        node = column_node->getExpression();
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Storages/StorageDistributed.cpp]: Do you mean to set `node->setAlias(column_node->getColumnName())` for replaced node? Added, let's see if it changes anything

File: src/Planner/CollectTableExpressionData.cpp (hunk 4 above)
Comment (vdimir) [reply src/Planner/CollectTableExpressionData.cpp:162]: In other cases, when we keep ColumnNode with expression and compute after `read`. Added it to comment

**Comment by vdimir** [comment]
> Is it possible to remove alias columns computation in StorageMerge?

I'm afraid there's not straightforward way to remove it, because planning of reading from source tables for `Merge` is hidden inside `read` method. But for Planner it's a blackbox and we could compute alias columns if StorageMerge itself had them. So, planner is not aware about underlying ALIAS columns and we still have to keep code to handle them in StorageMerge

**Comment by vdimir** [comment]
Test failures related to https://github.com/ClickHouse/ClickHouse/pull/60373

File: src/Planner/PlannerJoinTree.cpp [hunk 6]
```diff
@@ -864,6 +862,28 @@ JoinTreeQueryPlan buildQueryPlanForTableExpression(QueryTreeNodePtr table_expres
                     max_block_size,
                     max_streams);
 
+                const auto & alias_column_expressions = table_expression_data.getAliasColumnExpressions();
+                if (!alias_column_expressions.empty() && query_plan.isInitialized() && from_stage == QueryProcessingStage::FetchColumns)
+                {
+                    ActionsDAGPtr merged_alias_columns_actions_dag = std::make_shared<ActionsDAG>(query_plan.getCurrentDataStream().header.getColumnsWithTypeAndName());
+                    ActionsDAG::NodeRawConstPtrs action_dag_outputs = merged_alias_columns_actions_dag->getInputs();
+
+                    for (const auto & [column_name, alias_column_actions_dag] : alias_column_expressions)
```
Comment (filimonov) [inline src/Planner/PlannerJoinTree.cpp:871]: iteration over unordered map here. It makes the order of column in the resulting header unstable, and leads to columns swapping in the distributed queries.

Fixed in https://github.com/ClickHouse/ClickHouse/pull/94644

**Pull request merged**